File: backend/app/models/materials.py
"""Learning material chain: Material (logical resource) -> Document (parsed
representation) -> DocumentChunk (RAG unit with pgvector embedding) +
DocumentImage (extracted embedded raster with page provenance).

Only storage references live here — PDF binaries and image renditions go to
Neon Object Storage (prod) or local storage (dev), never PostgreSQL. One
Material has at most one Document (unique); the relationship is shaped so
document versioning can evolve later.
Image binaries are content-addressed by SHA-256 (duplicates share one object);
every page occurrence keeps its own row so provenance is never lost.
"""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import TYPE_CHECKING, Any

import sqlalchemy as sa
from pgvector.sqlalchemy import Vector
from sqlalchemy import JSON
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base
from app.models.enums import ExtractionMethod, MaterialStatus, MaterialType
from app.models.mixins import EMBEDDING_DIMENSIONS, FlexibleJSON, TimestampMixin

if TYPE_CHECKING:
    from app.models.learning import Project


class Material(Base, TimestampMixin):
    __tablename__ = "materials"

    id: Mapped[uuid.UUID] = mapped_column(sa.Uuid, primary_key=True, default=uuid.uuid4)
    project_id: Mapped[uuid.UUID] = mapped_column(
        sa.ForeignKey("projects.id", ondelete="CASCADE"), nullable=False, index=True
    )
    name: Mapped[str] = mapped_column(sa.String(200), nullable=False)
    type: Mapped[MaterialType] = mapped_column(
        sa.Enum(MaterialType, name="material_type"), nullable=False, default=MaterialType.PDF
    )
    description: Mapped[str | None] = mapped_column(sa.Text, nullable=True)
    status: Mapped[MaterialStatus] = mapped_column(
        sa.Enum(MaterialStatus, name="material_status"),
        nullable=False,
        default=MaterialStatus.QUEUED,
        index=True,
    )
    storage_provider: Mapped[str] = mapped_column(sa.String(32), nullable=False, default="local")
    storage_key: Mapped[str | None] = mapped_column(sa.String(512), nullable=True)
    original_filename: Mapped[str | None] = mapped_column(sa.String(255), nullable=True)
    mime_type: Mapped[str | None] = mapped_column(sa.String(128), nullable=True)
    file_size: Mapped[int | None] = mapped_column(sa.BigInteger, nullable=True)
    checksum: Mapped[str | None] = mapped_column(sa.String(128), nullable=True)
    processing_started_at: Mapped[datetime | None] = mapped_column(
        sa.DateTime(timezone=True), nullable=True
    )
    processing_completed_at: Mapped[datetime | None] = mapped_column(
        sa.DateTime(timezone=True), nullable=True
    )
    processing_error: Mapped[str | None] = mapped_column(sa.Text, nullable=True)
    retry_count: Mapped[int] = mapped_column(sa.Integer, nullable=False, default=0)
    archived_at: Mapped[datetime | None] = mapped_column(sa.DateTime(timezone=True), nullable=True)

    __table_args__ = (
        sa.CheckConstraint("file_size IS NULL OR file_size >= 0", name="ck_materials_file_size"),
        sa.CheckConstraint("retry_count >= 0", name="ck_materials_retry_count"),
        # Anchor for DocumentImage composite FKs (same pattern as
        # QuestionConcept): image links resolve through their own project.
        sa.UniqueConstraint("project_id", "id", name="uq_materials_project_id"),
    )

    project: Mapped[Project] = relationship(back_populates="materials")
    document: Mapped[Document | None] = relationship(
        back_populates="material",
        cascade="all, delete-orphan",
        passive_deletes=True,
        uselist=False,
    )


class Document(Base, TimestampMixin):
    __tablename__ = "documents"

    id: Mapped[uuid.UUID] = mapped_column(sa.Uuid, primary_key=True, default=uuid.uuid4)
    material_id: Mapped[uuid.UUID] = mapped_column(
        sa.ForeignKey("materials.id", ondelete="CASCADE"), nullable=False, unique=True
    )
    project_id: Mapped[uuid.UUID] = mapped_column(
        sa.ForeignKey("projects.id", ondelete="CASCADE"), nullable=False, index=True
    )
    page_count: Mapped[int | None] = mapped_column(sa.Integer, nullable=True)
    extraction_method: Mapped[ExtractionMethod | None] = mapped_column(
        sa.Enum(ExtractionMethod, name="extraction_method"), nullable=True
    )
    language: Mapped[str | None] = mapped_column(sa.String(16), nullable=True)
    doc_metadata: Mapped[dict[str, Any] | None] = mapped_column(
        "metadata", FlexibleJSON, nullable=True
    )

    __table_args__ = (
        sa.CheckConstraint("page_count IS NULL OR page_count >= 0", name="ck_documents_pages"),
        sa.UniqueConstraint("project_id", "id", name="uq_documents_project_id"),
    )

    material: Mapped[Material] = relationship(back_populates="document")
    project: Mapped[Project] = relationship(back_populates="documents")
    chunks: Mapped[list[DocumentChunk]] = relationship(
        back_populates="document", cascade="all, delete-orphan", passive_deletes=True
    )
    images: Mapped[list[DocumentImage]] = relationship(
        back_populates="document",
        cascade="all, delete-orphan",
        passive_deletes=True,
        foreign_keys="DocumentImage.document_id",
    )


class DocumentChunk(Base, TimestampMixin):
    """RAG retrieval unit. ``embedding`` is 768-d (Google gemini-embedding-001,
    truncated via output_dimensionality).

    Vector search strategy: exact (brute-force) scan scoped by ``project_id``.
    At assignment scale this is faster and simpler than an approximate index, which
    pgvector only recommends past ~10k+ vectors and which would require maintenance
    tuning. When chunks exceed ~100k rows, add an HNSW index with
    ``vector_cosine_ops`` in a new migration. See docs/DATABASE.md.
    """

    __tablename__ = "document_chunks"

    id: Mapped[uuid.UUID] = mapped_column(sa.Uuid, primary_key=True, default=uuid.uuid4)
    document_id: Mapped[uuid.UUID] = mapped_column(
        sa.ForeignKey("documents.id", ondelete="CASCADE"), nullable=False, index=True
    )
    project_id: Mapped[uuid.UUID] = mapped_column(
        sa.ForeignKey("projects.id", ondelete="CASCADE"), nullable=False, index=True
    )
    chunk_index: Mapped[int] = mapped_column(sa.Integer, nullable=False)
    content: Mapped[str] = mapped_column(sa.Text, nullable=False)
    page_start: Mapped[int | None] = mapped_column(sa.Integer, nullable=True)
    page_end: Mapped[int | None] = mapped_column(sa.Integer, nullable=True)
    section_title: Mapped[str | None] = mapped_column(sa.String(255), nullable=True)
    token_count: Mapped[int | None] = mapped_column(sa.Integer, nullable=True)
    embedding: Mapped[list[float] | None] = mapped_column(
        # none_as_null keeps SQLite semantics identical to PostgreSQL (Python
        # None <-> SQL NULL), so IS NULL / IS NOT NULL filters behave the same
        # in tests as they do against pgvector in production.
        Vector(EMBEDDING_DIMENSIONS).with_variant(JSON(none_as_null=True), "sqlite"),
        nullable=True,
    )
    chunk_metadata: Mapped[dict[str, Any] | None] = mapped_column(
        "metadata", FlexibleJSON, nullable=True
    )

    __table_args__ = (
        sa.UniqueConstraint("document_id", "chunk_index", name="uq_chunks_document_index"),
        sa.CheckConstraint("chunk_index >= 0", name="ck_chunks_index"),
        sa.CheckConstraint("page_start IS NULL OR page_start >= 0", name="ck_chunks_page_start"),
        sa.CheckConstraint("token_count IS NULL OR token_count >= 0", name="ck_chunks_tokens"),
    )

    document: Mapped[Document] = relationship(back_populates="chunks")
    project: Mapped[Project] = relationship(back_populates="chunks")


class DocumentImage(Base, TimestampMixin):
    """One embedded-raster occurrence on a document page. Binaries live in
    storage under a SHA-256 content key (duplicate binaries share one object);
    every page placement keeps its own row so provenance is exact.

    Project isolation is enforced at the database level: the composite FKs
    guarantee ``Material.project_id == Document.project_id ==
    DocumentImage.project_id`` (same pattern as ``QuestionConcept``). The
    older single-column FKs are retained; the composite ones are authoritative.
    """

    __tablename__ = "document_images"

    id: Mapped[uuid.UUID] = mapped_column(sa.Uuid, primary_key=True, default=uuid.uuid4)
    document_id: Mapped[uuid.UUID] = mapped_column(
        sa.ForeignKey("documents.id", ondelete="CASCADE"), nullable=False, index=True
    )
    material_id: Mapped[uuid.UUID] = mapped_column(
        sa.ForeignKey("materials.id", ondelete="CASCADE"), nullable=False, index=True
    )
    project_id: Mapped[uuid.UUID] = mapped_column(
        sa.ForeignKey("projects.id", ondelete="CASCADE"), nullable=False, index=True
    )
    page_number: Mapped[int] = mapped_column(sa.Integer, nullable=False)
    image_index: Mapped[int] = mapped_column(sa.Integer, nullable=False)
    width: Mapped[int] = mapped_column(sa.Integer, nullable=False)
    height: Mapped[int] = mapped_column(sa.Integer, nullable=False)
    mime_type: Mapped[str] = mapped_column(sa.String(64), nullable=False)
    file_size: Mapped[int] = mapped_column(sa.BigInteger, nullable=False, default=0)
    storage_key: Mapped[str] = mapped_column(sa.String(512), nullable=False)
    sha256: Mapped[str] = mapped_column(sa.String(64), nullable=False, index=True)

    __table_args__ = (
        sa.ForeignKeyConstraint(
            ["project_id", "document_id"],
            ["documents.project_id", "documents.id"],
            ondelete="CASCADE",
            name="fk_docimg_document_in_project",
        ),
        sa.ForeignKeyConstraint(
            ["project_id", "material_id"],
            ["materials.project_id", "materials.id"],
            ondelete="CASCADE",
            name="fk_docimg_material_in_project",
        ),
        sa.UniqueConstraint(
            "document_id", "page_number", "image_index", name="uq_docimg_document_page_index"
        ),
        sa.Index("ix_docimg_document_page", "document_id", "page_number"),
        sa.CheckConstraint("page_number >= 1", name="ck_docimg_page"),
        sa.CheckConstraint("image_index >= 0", name="ck_docimg_index"),
        sa.CheckConstraint("width >= 1", name="ck_docimg_width"),
        sa.CheckConstraint("height >= 1", name="ck_docimg_height"),
        sa.CheckConstraint("file_size >= 0", name="ck_docimg_file_size"),
    )

    document: Mapped[Document] = relationship(back_populates="images", foreign_keys=[document_id])

File: frontend/src/api/materials.ts
import { apiClient } from "./client";
import type {
  DocumentChunk,
  DocumentImage,
  KnowledgeStatus,
  Material,
  MaterialDetail,
  MaterialDocument,
  MaterialStatus,
  Page,
} from "@/types";

interface BackendMaterial {
  id: string;
  project_id: string;
  name: string;
  type: string;
  status: MaterialStatus;
  storage_provider: string;
  original_filename: string | null;
  mime_type: string | null;
  file_size: number | null;
  processing_error: string | null;
  retry_count: number;
  processing_started_at: string | null;
  processing_completed_at: string | null;
  page_count?: number | null;
  chunk_count?: number | null;
  created_at: string;
  updated_at: string;
}

interface BackendDocument {
  id: string;
  material_id: string;
  project_id: string;
  page_count: number | null;
  extraction_method: string | null;
  language: string | null;
  doc_metadata?: Record<string, unknown> | null;
  chunk_count: number;
  image_count: number;
  created_at: string;
  updated_at: string;
}

interface BackendImage {
  id: string;
  document_id: string;
  page_number: number;
  image_index: number;
  width: number;
  height: number;
  mime_type: string;
  file_size: number;
}

interface BackendChunk {
  id: string;
  document_id: string;
  project_id: string;
  chunk_index: number;
  content: string;
  page_start: number | null;
  page_end: number | null;
  section_title: string | null;
  chunk_metadata?: Record<string, unknown> | null;
}

function toMaterial(m: BackendMaterial): Material {
  return {
    id: m.id,
    projectId: m.project_id,
    name: m.name,
    type: m.type,
    status: m.status,
    originalFilename: m.original_filename,
    mimeType: m.mime_type,
    fileSize: m.file_size,
    processingError: m.processing_error,
    retryCount: m.retry_count,
    pageCount: m.page_count ?? null,
    chunkCount: m.chunk_count ?? null,
    createdAt: m.created_at,
    updatedAt: m.updated_at,
  };
}

function toDocument(d: BackendDocument): MaterialDocument {
  return {
    id: d.id,
    materialId: d.material_id,
    projectId: d.project_id,
    pageCount: d.page_count,
    extractionMethod: d.extraction_method,
    language: d.language,
    chunkCount: d.chunk_count,
    imageCount: d.image_count ?? 0,
  };
}

function toImage(i: BackendImage): DocumentImage {
  return {
    id: i.id,
    documentId: i.document_id,
    pageNumber: i.page_number,
    imageIndex: i.image_index,
    width: i.width,
    height: i.height,
    mimeType: i.mime_type,
    fileSize: i.file_size,
  };
}

function toChunk(c: BackendChunk): DocumentChunk {
  return {
    id: c.id,
    documentId: c.document_id,
    projectId: c.project_id,
    chunkIndex: c.chunk_index,
    content: c.content,
    pageStart: c.page_start,
    pageEnd: c.page_end,
  };
}

interface BackendMaterialDetail extends BackendMaterial {
  document: BackendDocument | null;
  image_count: number;
  knowledge?: {
    status: string;
    embedded: number;
    total: number;
    image_count: number;
    images_by_page: Record<string, number>;
  } | null;
}

export const materialsApi = {
  async list(projectId: string, params: { page?: number; pageSize?: number } = {}) {
    const res = await apiClient.get<Page<BackendMaterial>>(
      `/api/v1/projects/${projectId}/materials`,
      { params: { page: params.page ?? 1, page_size: params.pageSize ?? 20 } },
    );
    return { ...res.data, items: res.data.items.map(toMaterial) };
  },

  async upload(
    projectId: string,
    file: File,
    title: string,
    onProgress?: (percent: number) => void,
  ): Promise<Material> {
    const form = new FormData();
    form.append("file", file, file.name);
    if (title.trim()) form.append("title", title.trim());
    const res = await apiClient.post<BackendMaterial>(
      `/api/v1/projects/${projectId}/materials`,
      form,
      {
        headers: { "Content-Type": "multipart/form-data" },
        onUploadProgress: (e) => {
          if (e.total) onProgress?.(Math.round((e.loaded / e.total) * 100));
        },
      },
    );
    return toMaterial(res.data);
  },

  async get(projectId: string, materialId: string): Promise<MaterialDetail> {
    const res = await apiClient.get<BackendMaterialDetail>(
      `/api/v1/projects/${projectId}/materials/${materialId}`,
    );
    return {
      ...toMaterial(res.data),
      imageCount: res.data.image_count ?? 0,
      document: res.data.document ? toDocument(res.data.document) : null,
      knowledge: res.data.knowledge
        ? {
            status: res.data.knowledge.status as KnowledgeStatus,
            embedded: res.data.knowledge.embedded,
            total: res.data.knowledge.total,
            imageCount: res.data.knowledge.image_count ?? 0,
            imagesByPage: res.data.knowledge.images_by_page ?? {},
          }
        : null,
    };
  },

  async images(projectId: string, materialId: string): Promise<DocumentImage[]> {
    const res = await apiClient.get<Page<BackendImage>>(
      `/api/v1/projects/${projectId}/materials/${materialId}/images`,
      { params: { page_size: 100 } },
    );
    return res.data.items.map(toImage);
  },

  async imageBlob(projectId: string, materialId: string, imageId: string): Promise<Blob> {
    const res = await apiClient.get(
      `/api/v1/projects/${projectId}/materials/${materialId}/images/${imageId}/content`,
      { responseType: "blob" },
    );
    return res.data as Blob;
  },

  async pdfBlob(projectId: string, materialId: string): Promise<Blob> {
    const res = await apiClient.get(
      `/api/v1/projects/${projectId}/materials/${materialId}/content`,
      { responseType: "blob" },
    );
    return res.data as Blob;
  },

  async chunks(
    projectId: string,
    materialId: string,
    params: { page?: number; pageSize?: number } = {},
  ) {
    const res = await apiClient.get<Page<BackendChunk>>(
      `/api/v1/projects/${projectId}/materials/${materialId}/document/chunks`,
      { params: { page: params.page ?? 1, page_size: params.pageSize ?? 10 } },
    );
    return { ...res.data, items: res.data.items.map(toChunk) };
  },

  async reprocess(projectId: string, materialId: string): Promise<Material> {
    const res = await apiClient.post<BackendMaterial>(
      `/api/v1/projects/${projectId}/materials/${materialId}/reprocess`,
    );
    return toMaterial(res.data);
  },

  async archive(projectId: string, materialId: string): Promise<Material> {
    const res = await apiClient.delete<BackendMaterial>(
      `/api/v1/projects/${projectId}/materials/${materialId}`,
    );
    return toMaterial(res.data);
  },
};
